pantalaimon/daemon.py

```diff
@@ -1039,48 +1107,74 @@ async def search(self, request):
 
         return web.json_response(result, headers=CORS_HEADERS, status=200)
 
-    async def download(self, request):
-        server_name = request.match_info["server_name"]
-        media_id = request.match_info["media_id"]
-        file_name = request.match_info.get("file_name")
+    async def upload(self, request):
+        file_name = request.query.get("filename", "")
+        content_type = request.headers.get("Content-Type", "application/octet-stream")
+        client = next(iter(self.pan_clients.values()))
+
+        body = await request.read()
+        try:
+            response, maybe_keys = await client.upload(
+                data_provider=BufferedReader(BytesIO(body)),
+                content_type=content_type,
+                filename=file_name,
+                encrypt=True,
+            )
+
+            if not isinstance(response, UploadResponse):
+                return web.Response(
+                    status=response.transport_response.status,
+                    content_type=response.transport_response.content_type,
+                    headers=CORS_HEADERS,
+                    body=await response.transport_response.read(),
+                )
+
+            self.store.save_upload(self.name, response.content_uri, maybe_keys)
```

Comment:
we upload the encrypted media as `UploadInfo`, then whenever we intercept the url for that upload we save the related `MediaInfo`: https://github.com/matrix-org/pantalaimon/pull/79/files#diff-bffd867b07847bef716053411bb062ba167eaf511b8a36df42a6f53f01fa4b34R856

Reply:
I was a bit confused what `_map_media_info()` is supposed to do because it stores the media info but it's only used when we re-upload the file as a non-encrypted one. Seems like it is indeed used to reupload encrypted files as non-encrypted ones.

So the problem here is that `MediaInfo` gets only populated for non-encrypted rooms, thus if a client uploads a file to an encrypted room and later on tries to download it the `MediaInfo` will be missing.

In code this means that `_map_media_upload` is only called [here](https://github.com/matrix-org/pantalaimon/pull/79/files#diff-bffd867b07847bef716053411bb062ba167eaf511b8a36df42a6f53f01fa4b34R923) and you can see that it's inside a `if not encrypt` branch.

If the media info got saved when we did that upload this omission wouldn't matter so I do think it's the better way to handle this. You can just modify `save_upload()` to save the `MediaInfo` at the same time it saves the content uri and mime type of the upload.

Reply:
I see the problem now
but I've already pushed changes that will store `MediaInfo` also for the crypted messages :)

please, let me know if you are fine with it or you want me to move on the upload part

Reply:
Yes please, otherwise we're storing the stuff twice in the db and we would probably want to implement deleting uploads. This way we can just ignore uploads and leave them be forever.